src/pdf/filter.rs

```diff
@@ -0,0 +1,111 @@
+use mupdf_sys::*;
+
+use std::ffi::CStr;
+use std::os::raw::{c_char, c_void};
+use std::ptr;
+
+use crate::pdf::PdfDocument;
+use crate::{Image, Matrix, Rect};
+
+#[derive(Clone, Copy)]
+pub struct PdfFilterOptions {
+    pub(crate) inner: *mut pdf_filter_options,
+}
+
+// Callback types
+// TODO: I don't think this is necessary for anything.
+pub type ImageFilter = fn(ctm: &Matrix, name: &str, image: &Image) -> Image;
+pub type TextFilter = fn(ucsbuf: i32, ucslen: i32, trm: &Matrix, ctm: &Matrix, bbox: &Rect) -> i32;
+pub type AfterTextObject = fn(doc: &PdfDocument, chain: &pdf_processor, ctm: &Matrix);
+pub type EndPage = fn();
+
+impl PdfFilterOptions {
+    pub(crate) unsafe fn from_raw(ptr: *mut pdf_filter_options) -> Self {
+        Self { inner: ptr }
+    }
+
+    pub fn ascii(&self) -> bool {
+        unsafe { (*self.inner).ascii != 0 }
+    }
+
+    pub fn set_ascii(&mut self, value: bool) -> &mut Self {
+        unsafe {
+            (*self.inner).ascii = if value { 1 } else { 0 };
+        }
+        self
+    }
+
+    pub fn recurse(&self) -> bool {
+        unsafe { (*self.inner).recurse != 0 }
+    }
+
+    pub fn set_recurse(&mut self, value: bool) -> &mut Self {
+        unsafe {
+            (*self.inner).recurse = if value { 1 } else { 0 };
+        }
+        self
+    }
+
+    pub fn sanitize(&self) -> bool {
+        unsafe { (*self.inner).sanitize != 0 }
+    }
+
+    pub fn set_sanitize(&mut self, value: bool) -> &mut Self {
+        unsafe {
+            (*self.inner).sanitize = if value { 1 } else { 0 };
+        }
+        self
+    }
+
+    pub fn instance_forms(&self) -> bool {
+        unsafe { (*self.inner).instance_forms != 0 }
+    }
+
+    pub fn set_instance_forms(&mut self, value: bool) -> &mut Self {
+        unsafe {
+            (*self.inner).instance_forms = if value { 1 } else { 0 };
+        }
+        self
+    }
+
+    pub fn set_image_filter(
+        &mut self,
+        wrapper: fn(&Matrix, &str, &Image) -> Option<Image>,
+    ) -> &mut Self {
+        use once_cell::sync::OnceCell;
+
+        static WRAPPER: OnceCell<fn(&Matrix, &str, &Image) -> Option<Image>> = OnceCell::new();
+        // TODO: for some reason I can't use unwrap here??
+        let _ = WRAPPER.set(wrapper);
```

Comment:
Can we store the wrapper fn in `PdfFilterOptions` itself? For example:

```rust
#[derive(Clone, Copy)]
pub struct PdfFilterOptions {
    pub(crate) inner: *mut pdf_filter_options,
    image_filter: Option< ImageFilter>,
}
```

Reply:
The thing is that in order to access the filter inside the function it needs to be a static variable. So in any case I could save `image_filter` into the struct, but when the function is called I would still need the `OnceCell`/`static mut` initialization.

Reply:
OK, I need to take deeper look and try things out when I'm free.